rfcs/2020-10-29-3736-remap-chain-parsing.md

```diff
@@ -0,0 +1,181 @@
+no code blocks are run.
+
+The script is able to make the most of any processing that has been done to run the condition - for example, if a regular expression
+has been run, the block that is subsequently run can have access to the results of this regex. The Remap syntax also allows assignment to variables within 
+the condition. A condition is considered to be successful if the value it returns is not `false` or `nil`. This allows a regular expression match to pass 
+the condition and pass it's results into the block via the variable assignment.
+
+The following code will be possible:
+
+```coffee
+if $match = matches(.message, /^Started (?P<method>[^\s]*) for (?P<remote_addr>[^\s]*)/) {
+  .method = $match.method
+  .remote_addr = $match.remote_addr
+  .source = "nginx"
+} else if $match = matches(.message, /^(?P<remote_addr>[^\s]*).*"(?P<method>[^\s]*).*"$/ {
+  .method = $match.method
+  .remote_addr = $match.remote_addr
+  .source = "haproxy"
+} else {
+  .source = "none"
+}
+```
+
+*Note, `matches` is currently unimplemented, but in this example it is intended as a function that would match a regular expression and return any matching
+groups in a `Value::map`. If there were no match, it would return either `false` or `null`.* 
+
+
+
+## Rationale
+
+If this functionality can be encapsulated within the Remap language it will allow users to simplify their configuration. All processing
+involved would be contained within a single transform that can be precisely designed to do what the user needs in the minimum number
+of steps.
+
+## Prior Art
+
+Currently, we have the [Swimlanes](https://vector.dev/docs/reference/transforms/swimlanes/) transform. Swimlanes can route the event to a transform or sink
+according to a provided condition. These transforms could be other remap transforms that can do the required processing for that condition.
+
+Whilst this is possible, it does complicate the user's configuration as they will need multiple transforms specified. With this RFC in place, only
+one transform will likely be necessary to do the processing required.
+
+
+## Drawbacks
+
+Allowing `if` statements to work with non-boolean values is a controversial topic in programming language design. It is possible to create
+unintended bugs because the code hasn't been explicit enough about what should and should not pass the condition. Decisions need to be 
+made as to what actually constitutes a failure condition - for example, should a result of `0` be a pass or a fail? This does add additional
+documentation requirements and cognitive load when writing the script.
+
+The same can be said for allowing assignment within an `if` statement. The condition will be valid for both a single (`=`) and double (`==`) equals
+in the condition, yet both variations do very different things:
+
+```
+if .foo == true {
+
+}
+```
+
+```
+if .foo = true {
+
+}
+```
```

Comment:
This can be avoided by requiring non-boolean-operations to be wrapped in a group. See my comment on line ~177 for an example.